dnadoh/synthesize.py

```diff
@@ -0,0 +1,335 @@
+    marker = random.choice(list(candidates))
+    for person in people:
+        if person.genome[loc] == marker:
+            func(person)
+
+
+def adjust_weight(person):
+    """Adjust a person's weight upward by 10%."""
+    person.weight = _truncate(1.1 * person.weight, 1)
+
+
+# --------------------------------------------------------------------------------------
+# I/O
+# --------------------------------------------------------------------------------------
+
+
+def _write(output_stem, genomes, people):
+    """Write people as CSV."""
+    assert people
+    _write_overall(output_stem, people)
+    _write_reference_genome(output_stem, genomes)
+    _write_variants(output_stem, genomes, people)
+    _write_people(output_stem, people)
+
+
+def _write_overall(output_stem, people):
+    """Write DNA sequences and people for reference."""
+    filename = f"{output_stem}-overall.csv"
+    headings = people[0].dict().keys()
+    with open(filename, "w") as raw:
+        writer = csv.DictWriter(raw, fieldnames=headings)
+        writer.writeheader()
+        for person in people:
+            writer.writerow(person.dict())
+
+
+def _write_reference_genome(output_stem, genomes):
+    """Save the reference genome and related information."""
+    filename = f"{output_stem}-reference.json"
+    data = {"genome": genomes.reference, "locations": list(sorted(genomes.locations))}
+    with open(filename, "w") as writer:
+        json.dump(data, writer)
+
+
+def _write_variants(output_stem, genomes, people):
+    """Write one variant file per person."""
+    width = len(str(len(people)))
+    for (pid, person) in enumerate(people):
+        pid_str = str(pid).zfill(width)
+        filename = f"{output_stem}-pid{pid_str}.csv"
+        with open(filename, "w") as raw:
+            writer = csv.DictWriter(raw, fieldnames=["loc", "base"])
+            writer.writeheader()
+            for i in range(len(genomes.reference)):
+                if person.genome[i] != genomes.reference[i]:
+                    writer.writerow({"loc": i, "base": person.genome[i]})
+
+
+def _write_people(output_stem, people):
+    """Write phenotypic data for people."""
+    filename = f"{output_stem}-people.csv"
+    headings = people[0].dict()
+    del headings["genome"]
+    headings["pid"] = 0
+    with open(filename, "w") as raw:
+        writer = csv.DictWriter(raw, fieldnames=headings)
+        writer.writeheader()
+        for (pid, person) in enumerate(people):
+            details = person.dict()
+            del details["genome"]
+            details["pid"] = pid
+            writer.writerow(details)
+
+
+# --------------------------------------------------------------------------------------
+# Utilities
+# --------------------------------------------------------------------------------------
+
+
+def _other_bases(seq, loc):
+    """Create a list of bases minus the one in the sequence at that location."""
+    return list(set(DNA) - {seq[loc]})
```

Comment:
why do you make this a list instead of keeping it a set?

Reply:
Because the result is used as an argument to `random.choice` which requires an ordered sequence (like a list). We could return a set here and convert to a list where needed, but there are two calls, so I figured it was cleaner to do the conversion once. You're right, it's far from obvious.